netCDF4/_netCDF4.pyx

```diff
@@ -1060,6 +1061,62 @@ Note that there is currently no support for mapping numpy structured arrays with
 unicode elements (dtype `U#`) onto netCDF compound types, nor is there support 
 for netCDF compound types with vlen string components.
 
+## <div id='section15'>15) In-memory (diskless) Datasets.
+
+You can create netCDF Datasets whose content is held in memory
+instead of in a disk file.  There are two ways to do this.  If you
+don't need access to the memory buffer containing the Dataset from
+within python, the best way is to use the `diskless=True` keyword
+argument when creating the Dataset.  If you want to save the Dataset
+to disk when you close it, also set `persist=True`.  If you want to
+create a new read-only Dataset from an existing python memory buffer, use the
+`memory` keyword argument to pass the memory buffer when creating the Dataset.
+If you want to create a new in-memory Dataset, and then access the memory buffer
+directly from Python, use the `memory` keyword argument to specify the
+estimated size of the Dataset in bytes when creating the Dataset with
+`mode='w'`.  Then, the `Dataset.close` method will return a python memory
+buffer representing the Dataset. Below are examples illustrating both
+approaches.
+
+    :::python
+    >>> # create a diskless (in-memory) Dataset, and persist the file
+    >>> # to disk when it is closed.
+    >>> nc = Dataset('diskless_example.nc','w',diskless=True,persist=True)
+    >>> nc.history = 'test of diskless file capability'
+    >>> d = nc.createDimension('x',None)
+    >>> v = nc.createVariable('v',np.int32,'x')
+    >>> v[0:5] = np.arange(5)
+    >>> print(nc)
+    >>> print(nc['v'][:])
+    >>> nc.close() # file saved to disk
+    >>> # create an in-memory dataset from an existing python memory
+    >>> # buffer.
+    >>> # read the newly created netcdf file into a python bytes object.
+    >>> f = open('diskless_example.nc', 'rb')
+    >>> nc_bytes = f.read(); f.close()
+    >>> # create a netCDF in-memory dataset from the bytes object.
+    >>> nc = Dataset('inmemory.nc', memory=nc_bytes)
+    >>> print(nc)
+    >>> print(nc['v'][:])
+    >>> nc.close()
+    >>> # create an in-memory Dataset and retrieve memory buffer
+    >>> # estimated size is 1028 bytes - this is actually only
+    >>> # used if format is NETCDF3 (ignored for NETCDF4/HDF5 files).
+    >>> nc = Dataset('inmemory.nc', mode='w',memory=1028)
+    >>> d = nc.createDimension('x',None)
+    >>> v = nc.createVariable('v',np.int32,'x')
+    >>> v[0:5] = np.arange(5)
+    >>> nc_buf = nc.close() # close returns memory buffer.
+    >>> # save nc_buf to disk, read it back in and check.
+    >>> # tobytes method of cython memory buffer converts to bytes.
+    >>> f = open('inmemory.nc', 'w')
+    >>> f.write(nc_buf.tobytes()); f.close()
```

Comment:
You can also just pass a buffer directly into the write() method.

Reply:
No -that doesn't work either, probably for the same reason that bytes(buffer) doesn't.

Reply:
actually, this does work for NETCDF3 files, but for NETCDF4 files it throws this error
    TypeError: expected a string or other character buffer object

Reply:
Dataset.close now returns a python memoryview, which can be passed directly to the write() method.  This also makes the MemBuf.tobytes() method redundant.